### ISSUE TITLE
Bang method equivalents

### ISSUE THREAD

**Issue by ginty** [Origen-SDK/o2#23]
Python does not permit bangs (!) and similar symbols in method names, so the O1 convention of appending a bang to a method name to make it take action will not be possible in O2:

~~~ruby
pin("tdi").drive(1)    # Configure the pin to drive a 1, but don't apply it
pin("tdi").drive!(1)   # Configure the pin to drive a 1 and apply it

reg("blah").write(0x1234)   # Write the value to the register, but don't apply it to the device
reg("blah").write!(0x1234)  # Write the value to the register and apply it to the device
~~~

How should we handle this in O2 with consistency across all similar APIs?

What about having it such that the APIs are not applied by default, then we have an `apply()` method:

~~~python
pin("tdi").drive(1) # Configure the pin to drive a 1, but don't apply it
pin("tdi").apply()  # Apply the current state (which for a pin is equivalent to tester.cycle()

reg("blah").write(0x1234) # Write the value to the register, but don't apply it to the device
reg("blah").apply()
~~~

It would be nice to also support a 1-liner I suppose, what about:

~~~python
# add _and_apply version of all methods:
pin("tdi").drive_and_apply(1) 
reg("blah").write_and_apply(0x1234)
~~~

or

~~~python
# Have all methods accept an optional apply argument last
pin("tdi").drive(1, True) 
reg("blah").write(0x1234, True)
~~~

Thoughts?

**Comment by ginty** [comment]
From @coreyeng's comment here https://github.com/Origen-SDK/o2/pull/21#issuecomment-564367914, we could have the all these methods return self (which is generally nice for chaining, O1 sometimes did that but it would be nice to be consistent here), then we can have the 1-liners:

~~~python
pin("tdi").drive(1).apply() 
reg("blah").write(0x1234).apply()
~~~

**Comment by ginty** [comment]
I just realized that the proposal for how to represent vectors internally (https://github.com/Origen-SDK/o2/issues/18) implies a slight change in behaviour with regards to applying pin state and `tester.cycle`.

In O1, `tester.cycle` meant "apply all pins and advance time by one cycle", whereas in O2 it will only mean "advance time by one cycle".

I think that is better and brings consistency between how reg and pin states are handled - in both cases they will not actually be applied to the DUT until `apply()` is called on them.

**Comment by info-rchitect** [comment]
@ginty I don't mind the apply if it is chained as it is absolutely clear in meaning.  I can also see the case for 'drive_and_apply' but perhaps syntactic sugar is round 2 after the API is fully up and running?

**Comment by ginty** [comment]
I think it should be either/or and we should avoid multiple ways of doing the same thing.
I think the chained approach has clear advantages on the implementation side and is elegant on the user side, so that's the one that gets my vote.

**Comment by info-rchitect** [comment]
> I think it should be either/or and we should avoid multiple ways of doing the same thing.
> I think the chained approach has clear advantages on the implementation side and is elegant on the user side, so that's the one that gets my vote.

Works for me.

**Comment by coreyeng** [comment]
> > I think it should be either/or and we should avoid multiple ways of doing the same thing.
> > I think the chained approach has clear advantages on the implementation side and is elegant on the user side, so that's the one that gets my vote.
> 
> Works for me.

Kinda funny to think of us doing a 180 from Ruby. I'm so used to throwing everything I can think of into an API, which Ruby loved but which Python frown upon.

Anyway, is something like `reg.write(0).apply()` much different than what I have for pins now, which would be applied to regs as `reg.posture(0).write()`? I went with the latter since its clear that the data in the register is consistent across actions. Thinking back when I was introducing O1 to users with not much programming experience, when we threw bit collections into the mix, some of them were getting a bit confused by things like:

~~~ruby
reg.bits[0].read!(1)
reg.bits[1].write!(1)
  #=> actually writes 0x3 since the data from the read above persists.
~~~

I recall having to explain that the state was consistent across read/write operations which when written back-to-back like above makes sense,, but when sprinkled out can get lost (again, thinking in terms of new users here). My equivalent would be:

~~~python
reg.bits[0].posture(1).read()
reg.bits[1].posture(1).write()
~~~

In what I have now, `posture` would be equivalent to a non-bang, and the non-bang Python method would be equivalent to the bang-method counterpart... I realize the confusion here as well. But I think setting a consistent state then kicking off an action is a clearer way... I may be biased though.

I think part of what might be confusing in #21 is that I represented a pin state as a bool, so I've got some conversion stuff running around, which may have implied its not a 0/1 value (which it is).

**Comment by coreyeng** [comment]
Though as I was writing the above, I realized a 'syntatic sugar' method would be to combine the two... which then gets us right back to where we were in O1, except without the bangs... so maybe I'm not really solving the problem here.

~~~python
reg.pin[0].posture(1).read()

# shortcut for the above
reg.pin[0].read(1) # ... same as what I was just complaining about
~~~

**Comment by ginty** [comment]
> Kinda funny to think of us doing a 180 from Ruby. I'm so used to throwing everything I can think of into an API, which Ruby loved but which Python frown upon.

Yeah, I agree.

I do think though that we did pay a significant cost for supporting that within the framework and a significant number of bugs were caused by it. So I can appreciate the benefits we will see in simplifying the framework implementation by being stricter on what is allowed by the application.

I do like the idea of having a single method name (`apply()`) being used in all cases to apply state to the DUT, rather than having different methods (`write()`, `read()`) as the final operation as you currently have it.

I also still think the `posture` API doesn't really do what your intention is. It doesn't really setup the fully postured state, only the data value. i.e. it has to be combined with a read or write call to fully establish "I want to read/write this pin and I expect the data to be x".

On the comments about user confusion, would it help if we separate out the data setting from the action, a hybrid of what I am suggesting from what you currently have (sorry, I still don't like posture if it only really means data):

~~~python
pin("tdi").set_data(1).drive().apply() 
reg("blah").set_data(0x1234).write().apply()
~~~

Does that help in cases like this:

~~~python
reg.bits[0].set_data(1).read().apply()
reg.bits[1].set_data(1).write().apply()
~~~

Maybe it makes it a bit clearer that the data setting is independent of the action type and therefore easier for a user to accept/expect that it will persist?

**Comment by coreyeng** [comment]
I think separating it out is clearer, and ultimately that's all I was trying to do with `posture`. I'm fine with the hybrid method. Its a bit more verbose but I think is very clear in what's going on. My worry would be someone later tries to merge the above (as @info-rchitect already tried with the syntatic sugar of `drive_and_apply`) which then gets us back to where we were before.

I like the hybrid case, but if the consensus is that its too verbose I think we should shift back to what you had originally.

**Comment by chrisnappi** [comment]
I think this looks good.  One thing that came up recently with O1 was the masking of pins in a bus.

So:
pin("porta").set_data(0x55).set_mask(0x0F).drive().apply() 
and
reg("blah").set_data(0x1234).set_mask(0xF).write().apply()

?

**Comment by ginty** [comment]
@chrisnappi, what effect would the mask have in those examples?

**Comment by chrisnappi** [comment]
Set lower nibble to don't care.

**Comment by coreyeng** [comment]
I think a problem with `set_mask` at that level is it's not clear that `write` won't be `or-ing` the data - a thing new users almost always mess up once or twice. I don't think we'd want to support a `set_mask` for the write operation at all. I think you'd be better off splicing in that case.

**Comment by chrisnappi** [comment]
Sorry, copy paste error, I meant:
pin("porta").set_data(0x55).set_mask(0x0F).compare().apply()
and
reg("blah").set_data(0x1234).set_mask(0xF).read().apply()

**Comment by ginty** [comment]
Should mask just be a argument to read/compare then:

~~~python
pin("porta").set_data(0x55).compare(mask=0xF).apply()
reg("blah").set_data(0x1234).read(mask=0xF).apply()
~~~

Which also implies to me that it would be non-sticky, is that ok?

**Comment by info-rchitect** [comment]
@ginty Will we support a sticky standalone mask function?

**Comment by chrisnappi** [comment]
I think that will handle most cases.  I do think we will want some syntax for setting don't care/hi-z at the port level as well though.  

For a register, I don't think it is common, but there are a few corner cases where register bits really can have an unknown value.

**Comment by coreyeng** [comment]
> I think that will handle most cases. I do think we will want some syntax for setting don't care/hi-z at the port level as well though.
> 
> For a register, I don't think it is common, but there are a few corner cases where register bits really can have an unknown value.

Agree, especially when we start integrating OrigenSim. Granted that's a ways away, but should keep that in mind.

**Comment by pderouen** [comment]
> 
> 
> I just realized that the proposal for how to represent vectors internally (#18) implies a slight change in behaviour with regards to applying pin state and `tester.cycle`.
> 
> In O1, `tester.cycle` meant "apply all pins and advance time by one cycle", whereas in O2 it will only mean "advance time by one cycle".
> 
> I think that is better and brings consistency between how reg and pin states are handled - in both cases they will not actually be applied to the DUT until `apply()` is called on them.

I'm having trouble picturing what use it would be to set a pin drive state, but not have it rendered as an action when time is advanced.

**Comment by ginty** [comment]
> I'm having trouble picturing what use it would be to set a pin drive state, but not have it rendered as an action when time is advanced.

Yeah, me either tbh.

An alternative could be to make `apply()` only a method for registers and just have all pin write/compare calls simply [generate vector actions](https://github.com/Origen-SDK/o2/issues/18) immediately.

Even if a pin changed state a few times before the next cycle it should come out in the wash and the last update should win.

Though we're now getting in the debate of whether even though `apply()` might not be that useful for pins, is it better to have it so that users can have the same mental model and expectation about how they work vs. registers?

**Comment by ginty** [comment]
Just copying in this comment from #22:

This seems like a pretty self-explanatory choice of primary action methods:

~~~python
reg.write()
reg.verify()
reg.capture()

pin.write()
pin.verify()
pin.capture()
~~~

**Comment by pderouen** [comment]
<code>apply()</code> could still be meaningful for pins if it is interpreted to mean advance 1 time unit. That's how we used the bang method.

**Comment by ginty** [comment]
I think this is potentially confusing then though:

~~~python
dut.pin("tdi").set_data(1).write()
dut.pin("tms").set_data(0).write().apply()

# Was my TDI value applied or not?!
~~~

I think the bang carried less baggage with it with regards to its potential meaning. Calling it `cycle()` and calling that on a pin object doesn't feel right to me either.

Having cycle/advance time as its own function call should feel natural to users I think and I've seen many other tools do it that way, even though its a bit more verbose vs O1.

So I think its either that we have the not-very-useful apply method do the same thing as for regs, and we keep it around for consistency:

~~~python
dut.pin("tdi").set_data(1).write().apply()
dut.pin("tms").set_data(0).write().apply()
tester.cycle()
~~~

Or else we just drop the `apply()` and make that a register-only concern:

~~~python
dut.pin("tdi").set_data(1).write()
dut.pin("tms").set_data(0).write()
tester.cycle()
~~~

Probably the latter for me.

**Comment by priyavadan** [comment]
to me for pins **dut.pin("tdi").drive(1)** is cleaner than **dut.pin("tdi").set_data(1).write()**

Similarly for registers, I would like to see something like **reg("reg1").set_mask(0xF).read(0x12345)**

It explicitly states
(1) I am driving pin TDI to 1
(2) I am setting a mask on reg1 and reading(0x12345)

**Comment by priyavadan** [comment]
as far as advancing time for pins, `tester.cycle()` is okay and i like @ginty 's suggestion of  dropping the `apply()` and making that a register-only concern.

**Comment by ginty** [comment]
@priyavadan, I think we need the concept of `.apply()` for the regs, do you agree or do you mean for the read to happen immediately in your example?

**Comment by priyavadan** [comment]
yes, we need the concept of `apply()` for registers. There are times we will be setting up the register values and then apply the code once later in the workflow. Sorry if I was unclear.

**Comment by coreyeng** [comment]
We'd need to be a bit careful with the return values for `read` or whatever if we add `set_mask` then. If `read` is kicking off the actual transaction then `reg.read(0x1234).set_mask(0xF)` wouldn't do the right thing.

We can still do that, just something to keep in mind. Are we also making `set_mask` sticky? If not, could use `with_mask` instead: `reg.with_mask(0xF).read(0x1234)`. I think this implies a non-sticky mask that is reset on the next read operation.

**Comment by priyavadan** [comment]
@coreyeng I like with_mask, it makes it clear that only the current read is being done with the mask in place

**Comment by redxeth** [comment]
Would `dut_apply()` be more clear than `apply()` ?  Basically without `dut_apply` anything is register model or pin model only.  Or potentially we just use `_bang` at the end of methods but to me that's too similar to the O1 style. What do you think.

**Comment by chrisnappi** [comment]
Don't like either of those options _especially_ not bang.  But I don't love apply either.  .execute ?

**Comment by redxeth** [comment]
Yeah.. I debated even putting it.  How about `make_it_so` ala Star Trek ;-)

**Comment by redxeth** [comment]
Or `engage` to be really nerdy
